### PR TITLE
[next] Add `enhanceApp` / `enhanceComponent`

### PR DESCRIPTION
**Pull request by elliottsj**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
  - https://github.com/zeit/next.js/tree/cf31021e25f8d90f91ef3f7cb4f36a46d52fc56d#customizing-renderpage
  - https://github.com/zeit/next.js/blob/7e7be0e2e8419b28d9c562e04cba1fc52a39f086/packages/next-server/server/render.tsx#L13-L32
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by jthegedus — APPROVED** [review]
lgtm, but another set of eyes would be nice

**Comment by armanio123** [comment]
@jthegedus I'll wait for another reviewer before merging; otherwise let me know if you feel is unnecessary and I can merge it right away.

**Review by joaovieira — APPROVED** [review]

**Comment by armanio123** [comment]
Great we got two reviewers. Merging.

**Pull request merged**
